pkg/processor/procjs/processor.go

```diff
@@ -110,75 +125,98 @@ func (p *Processor) jsContentRaw(goja.ConstructorCall) *goja.Object {
 	return p.runtime.ToValue(&r).ToObject(p.runtime)
 }
 
+func (p *Processor) jsContentStructured(goja.ConstructorCall) *goja.Object {
+	// TODO accept arguments
+	// We return a map[string]interface{} struct, however because we are
+	// not changing call.This instanceof will not work as expected.
+
+	r := make(map[string]interface{})
+	return p.runtime.ToValue(r).ToObject(p.runtime)
+}
+
 func (p *Processor) Process(_ context.Context, in record.Record) (record.Record, error) {
-	jsRecord := p.toJSRecord(in)
+	jsr := p.toJSRecord(in)
 
-	result, err := p.function(goja.Undefined(), jsRecord)
+	result, err := p.function(goja.Undefined(), jsr)
 	if err != nil {
 		return record.Record{}, cerrors.Errorf("failed to execute JS processor function: %w", err)
 	}
 
 	out, err := p.toInternalRecord(result)
+	if err == processor.ErrSkipRecord {
+		return record.Record{}, err
+	}
 	if err != nil {
 		return record.Record{}, cerrors.Errorf("failed to transform to internal record: %w", err)
 	}
 
-	// nil will be returned if the JS function has no return value at all
-	if out == nil {
-		return record.Record{}, processor.ErrSkipRecord
-	}
-
-	return *out, nil
+	return out, nil
 }
 
 func (p *Processor) toJSRecord(r record.Record) goja.Value {
-	// convert content to pointers to make it mutable
-	switch v := r.Payload.(type) {
-	case record.RawData:
-		r.Payload = &v
-	case record.StructuredData:
-		r.Payload = &v
+	extractData := func(d record.Data) interface{} {
+		switch v := d.(type) {
+		case record.RawData:
+			return &v
+		case record.StructuredData:
+			return map[string]interface{}(v)
+		}
+		return nil
 	}
 
-	switch v := r.Key.(type) {
-	case record.RawData:
-		r.Key = &v
-	case record.StructuredData:
-		r.Key = &v
+	jsr := jsRecord{
+		Position:  r.Position,
+		Operation: r.Operation.String(),
+		Metadata:  r.Metadata,
+		Key:       extractData(r.Key),
+		Payload: struct {
+			Before interface{}
+			After  interface{}
+		}{
```

Comment:
Do anonymous structs in Go create any overhead?

Reply:
Not that I know of, we'd have to measure it. Although I honestly doubt that it would make a difference in the JS processor which is already relatively slow itself.